tools/bridge/bridge/validation_utils.py

```diff
@@ -0,0 +1,73 @@
+from web3 import Web3
+from web3.contract import Contract
+from web3._utils.abi import abi_to_signature
+
+from bridge.contract_abis import MINIMAL_VALIDATOR_PROXY_ABI
+
+
+def validate_contract(contract: Contract) -> None:
+    """Verifies if a given contract exists on the chain.
+
+    It checks if any code is stored for the address. For events and functions
+    in the ABI it will be checked if their signature hash can be found within
+    the code. Throws an exception if the contract could not be verified. Else
+    return nothing.
+    """
+
+    contract_code = contract.web3.eth.getCode(contract.address)
+
+    if not contract_code:
+        raise ValueError(
+            f"The given contract address {contract.address} does not point to a contract!"
+        )
+
+    for description in contract.abi:
+        description_type = description.get("type", "function")
+
+        if description_type in ("constructor", "fallback"):
+            continue
+
+        signature = abi_to_signature(description)
+        signature_hash = Web3.keccak(text=signature)
+
+        if description_type == "function":
+            signature_hash = signature_hash[:4]
+
+        if signature_hash not in contract_code:
+            raise ValueError(
+                f"The contract at the given address {contract.address} does not"
+                f" support the {description_type} with the following signature: {signature}!"
+            )
+
+
+def validate_confirmation_permissions(home_bridge_contract: Contract, address: str):
```

Comment:
I don't think we should check this as it's (I think) impossible to do properly in light of the dynamic nature of the validator set:

- A validator might not be validator right now, but wants to confirm past pending transfers
- A validator might not be validator right now, but will be soon and wants to be ready as soon as the validator set changes

Reply:
> A validator might not be validator right now, but wants to confirm past pending transfers

 This does not work with the current implementation of the home bridge contract. The [require](https://github.com/trustlines-protocol/blockchain/blob/5a54d4105443fcc51eaf91fe1f9537b89f7d5253/contracts/contracts/bridge/HomeBridge.sol#L52) do a plain check for the current active set.

> A validator might not be validator right now, but will be soon and wants to be ready as soon as the validator set changes

This would lead to possibly many failing transactions until he become a validator. I think the disadvantages are higher than running the bridge client just a bit later. And possibly miss a transfer, because over 50% other validators made it to start the node earlier than you or synced faster.
Possible we could move this check somewhere else before a confirmation is made. But this seem to be quite verbose to me.

Reply:
Okay, in a call including Sascha we decided the following:
- have an initial check which log a warning if it fails
- check the permissions before each confirmation 

This allows the flexibility for valdiators to start a node earlier and longer than their actual active period.

Reply:
I implemented it. Please check if out and leave another review.